python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+```
+
+## Workflow Definitions
+
+A workflow is a class.
+
+Here is an example of a class-based workflow that calls a local activity to get
+a greeting with a signal to update the prefix and a query to get the latest
+greeting:
+
+```python
+import asyncio
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def create_greeting_activity(salutation: str, name: str) -> str:
+    return f"{salutation}, {name}!"
+
+@workflow.defn
+class GreetingWorkflow:
+    def __init__() -> None:
+        self._salutation = "Hello"
+        self._salutation_update = asyncio.Event()
+        self._complete = asyncio.Event()
+
+    @workflow.run
+    async def run(self, name: str) -> str:
+        while True:
+            # Store greeting (this is typed)
+            self._current_greeting = await workflow.execute_local_activity(
+                create_greeting_activity,
+                self._salutation,
+                name,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            workflow.logger.debug("Greeting set to %s", self._current_greeting)
+            
+            # Wait for salutation update or complete signal (this can be
+            # cancelled)
+            await asyncio.wait(
+                [self._salutation_update.wait(), self._complete.wait()],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            if self._salutation_update.is_set():
+                self._salutation_update.clear()
+            elif self._complete.is_set():
+                return self._current_greeting
+
+    @workflow.signal
+    async def update_salutation(self, salutation: str) -> None:
+        self._salutation = salutation
+        self._salutation_update.set()
+
+    @workflow.signal
+    async def complete_with_greeting(self) -> None:
+        self._complete.set()
+
+    @workflow.query
+    async def current_greeting(self) -> str:
+        return self._current_greeting
+
+def test_workflow(client: Client) -> None:
+    # Start the workflow
+    handle = await client.start_workflow(GreetingWorkflow.run, "John")
+
+    # Confirm the current greeting
+    assert "Hello, John!" == await handle.query(GreetingWorkflow.current_greeting)
+
+    # Change the salutation part
+    await handle.signal(GreetingWorkflow.update_salutation, "Hola")
+
+    # Now tell it to complete with the new greeting
+    await handle.signal(GreetingWorkflow.complete_with_greeting)
+
+    # Check result
+    assert "Hola, John!" == await handle.result()
+
+def create_worker(client: Client) -> Worker:
+    return Worker(
+      client=client,
+      activities=[create_greeting_activity],
+      workflows=[GreetingWorkflow],
+    )
+```
+
+Notes:
+
+* Workflows are classes
+  * Consideration was given to making workflows be functions (in addition to classes or having the class equivalent sit
```

Comment:
I've raised my concerns with this approach already but I'll just put them down here FTR:

1. With classes it's much harder to have reusable components because that requires multiple inheritance
2. With classes, because the start params are only provided to the `run` method, you cannot initialize your class in the constructor which in some cases requires making some attributes optional where they really are not resulting in redundant type assertions throughout the implementation.

Reply:
1. Yes, this is a limitation of class
2. There are ways we can make it a bit easier

Here are the concerns with the function approach:

1. You have to remember to manually set handlers everywhere - no good
2. The definition for your query/signal is divorced from the implementation, but your workflow gets to happily have the definition and implementation together - no good
3. You either don't get the benefits of class state or we have to support multiple ways of writing a workflow
4. For purposes of running (not yet determined how needed), functions referencing other functions are difficult to isolate without isolating the entire runtime. It can be hard to reason about isolated state when it's a loosely connected set of functions instead of a class we can ask they keep state within.
5. If workflows are functions, how do you define which queries/signals belong to which workflows at compile time? TS forces you mostly to file-per-workflow if you want to avoid such confusion which I don't think is a good model.

I can of course go back on this, but I might need to see some examples to justify it. For instance, two workflows with separate signals and queries and what their client calls look like.

Reply:
> 1. You have to remember to manually set handlers everywhere - no good

True for your class based proposal too

> 2. The definition for your query/signal is divorced from the implementation, but your workflow gets to happily have the definition and implementation together - no good

Some might say this is a good thing, especially since there are plenty of cases where the caller does not have the implementation available

> 3. You either don't get the benefits of class state or we have to support multiple ways of writing a workflow

It might not be too bad, I'm thinking of *also* adding class support to TS.

> 4. For purposes of running (not yet determined how needed), functions referencing other functions are difficult to isolate without isolating the entire runtime. It can be hard to reason about isolated state when it's a loosely connected set of functions instead of a class we can ask they keep state within.

It's easy enough create a class or access any variable in scope.
With functions one can test any part of their code where with classes there's extra setup involved.

> 5. If workflows are functions, how do you define which queries/signals belong to which workflows at compile time? TS forces you mostly to file-per-workflow if you want to avoid such confusion which I don't think is a good model.

TS is flexible and doesn't impose any restrictions on file structure, I've seen cases where signals are all defined in a central location and reuse in different workflows.

All of that being said, working with non local variables in Python is a bit inconvenient and classes might be a better fit here.

Reply:
> True for your class based proposal too

Not true, the definition and the handler are married. You can't create a definition without the handler like you can in TS.

> Some might say this is a good thing, especially since there are plenty of cases where the caller does not have the implementation available

I don't think this will be that common. And of course as mentioned before, having to keep two contracts in sync is rough.

> It might not be too bad, I'm thinking of also adding class support to TS.

During the last phase, I was constantly told not to have multiple definition/registration approaches.

> It's easy enough create a class or access any variable in scope. With functions one can test any part of their code where with classes there's extra setup involved.

I think this is an acceptable cost. If that setup is a real problem, I can only imagine the problems with the other things you have to setup for tests.

> TS is flexible and doesn't impose any restrictions on file structure, I've seen cases where signals are all defined in a central location and reuse in different workflows.

What this also means is you can have a signal that is not even on a workflow and no way in the type system to prevent using it. There is nothing at compile time in TS that puts signals on a workflow, and I think it's valuable to have such a thing easy to see for a dev. One could only imagine a complex workflow where I'd have to walk a bunch of code to see what signals are even on a workflow.

> All of that being said, working with non local variables in Python is a bit inconvenient and classes might be a better fit here.

In the same vein, functions are the Python code organization construct of choice and other similar Python solutions do let you decorate functions. Simply stated, "does the clarity of predefining multiple things belonging to a workflow outweigh the annoyance of a class?"

Reply:
I quite like the class based approach. I'm honestly not sure re-usable definitions/components actually net out to be a good thing in the long run. It seems very easy to break compatibility for some workflow you didn't realize your code was being used in. Stated differently: I think you want to keep the call depth of workflows shallow, because you want their execution path to be obvious. 

The way I might imagine having the function based approach work is by defining a workflow as a collection of signal/query definitions (which may come with implementations, or point to them) along with it's entry point signature/implementation, but I think as I brought that up in Slack there was no way to make that work statically with Python type checking.

Reply:
After taking a stab at functional workflows with Python, I'm now on "team classes".

Below is what the TS API would look like translated to Python.
The main issue I see with this approach is that Python doesn't have anonymous functions (it does have lambdas but those are quite limited).
This limitation requires extra typing which we can avoid with the class based approach.

```python
class State(dataclass):
  blocked: bool

unblock_signal = workflow.define_signal('unblock')

async def my_workflow(start_blocked: bool):
  state = State(blocked=start_blocked)

  @workflow.signal(unblock_signal)
  def unblock():  # Would have been better as an anonymous function
     state.blocked = False

  await workflow.wait_condition(lambda: not state.blocked)
  ...
```

To summarize, while there are advantages to the functional approach and a lot of the concerns of associating signals and queries with a workflow and verifying that those are handled can be done with the functional approach, it would be less convenient in Python, I'm dropping this issue.

Reply:
Technically you could have:

```python
@workflow.signal_defn
def unblock_signal(some_arg: str) -> None:
    ...

@workflow.defn
async def my_workflow() -> None:
    @unblock_signal
    def unblock(some_arg: str) -> None:
      doSomething()
```

And I think I could type it, but that's a lot of magic.